### PR TITLE
flamenco: spad allocation for accounts

### PR DESCRIPTION
**Pull request by ibhatt-jumptrading**

### CODE REVIEW

File: src/app/fdctl/run/tiles/fd_replay.c [hunk 1]
```diff
@@ -1675,6 +1682,19 @@ unprivileged_init( fd_topo_t *      topo,
...
+  /**********************************************************************/
+
+  /* TODO: The spad should probably have its own workspace. */
+  for( ulong i=0UL; i<tile->replay.tpool_thread_count; i++ ) {
+    ulong       total_mem_sz = fd_ulong_align_up( 128UL * FD_ACC_SZ_MAX, FD_SPAD_ALIGN );
+    uchar *     mem          = fd_wksp_alloc_laddr( ctx->blockstore_wksp, FD_SPAD_ALIGN, total_mem_sz, 999UL );
+    fd_spad_t * spad         = fd_spad_join( fd_spad_new( mem, total_mem_sz ) );
+    ctx->spads[ ctx->spad_cnt++ ] = spad;
+  }
+
```
Comment (ibhatt-jumptrading) [inline src/app/fdctl/run/tiles/fd_replay.c:1696]: nit: remove newline here

File: src/app/ledger/main.c [hunk 2]
```diff
@@ -423,6 +428,17 @@ fd_ledger_main_setup( fd_ledger_args_t * args ) {
   fd_bpf_scan_and_create_bpf_program_cache_entry( args->slot_ctx, args->slot_ctx->funk_txn, 1 );
   fd_funk_end_write( funk );
 
+  /* Allocate memory for the account scratch space */
+
+  args->spad_cnt = fd_tpool_worker_cnt( args->tpool );
+  for( ulong i=0UL; i<args->spad_cnt; i++ ) {
+    ulong       total_mem_sz = fd_ulong_align_up( 128UL * FD_ACC_SZ_MAX, FD_SPAD_ALIGN );
+    uchar *     mem          = fd_wksp_alloc_laddr( args->wksp, FD_SPAD_ALIGN, total_mem_sz, 999UL );
```
Comment (topointon-jump) [inline src/app/ledger/main.c:436]: can we add a comment explaining that this isn't thread-local memory, and probably should be in the future

File: src/flamenco/runtime/fd_runtime.c [hunk 3]
```diff
@@ -2067,24 +2062,27 @@ fd_runtime_execute_txns_in_waves_tpool( fd_exec_slot_ctx_t * slot_ctx,
                                         fd_capture_ctx_t *   capture_ctx,
                                         fd_txn_p_t *         all_txns,
                                         ulong                total_txn_cnt,
-                                        fd_tpool_t *         tpool ) {
+                                        fd_tpool_t *         tpool,
+                                        fd_spad_t * *        spads,
+                                        ulong                spad_cnt ) {
     int dump_txn = capture_ctx && slot_ctx->slot_bank.slot >= capture_ctx->dump_proto_start_slot && capture_ctx->dump_txn_to_pb;
     #define BATCH_SIZE (128UL)
+    ulong batch_size = fd_ulong_min( fd_tile_cnt(), BATCH_SIZE );
```
Comment (topointon-jump) [inline src/flamenco/runtime/fd_runtime.c:2070]: Explain where 128 came from (thread count)

File: src/flamenco/runtime/fd_runtime.c [hunk 4]
```diff
@@ -1066,6 +1067,12 @@ fd_txn_prep_and_exec_task( void  *tpool,
   fd_execute_txn_task_info_t * task_info = (fd_execute_txn_task_info_t *)tpool + m0;
   fd_exec_slot_ctx_t * slot_ctx = (fd_exec_slot_ctx_t *)args;
   // fd_capture_ctx_t * capture_ctx = (fd_capture_ctx_t *)reduce;
+    
+  ulong tile_idx = fd_tile_idx();
```
Comment (topointon-jump) [inline src/flamenco/runtime/fd_runtime.c:1071]: could we explain that there is a 1-1 mapping between the tiles and the tpool threads - there are no tiles that should not be "owned" by a tpool thread

File: src/flamenco/runtime/fd_executor.c [hunk 5]
```diff
@@ -1258,10 +1258,9 @@ fd_executor_setup_borrowed_accounts_for_txn( fd_exec_txn_ctx_t * txn_ctx ) {
     memcpy( borrowed_account->pubkey->key, acc, sizeof(fd_pubkey_t) );
 
     if( fd_txn_account_is_writable_idx( txn_ctx, (int)i ) ) {
-      void * borrowed_account_data = fd_valloc_malloc( txn_ctx->valloc, 8UL, fd_borrowed_account_raw_size( borrowed_account ) );
+      void * borrowed_account_data = fd_spad_alloc( txn_ctx->spad, FD_SPAD_ALIGN, FD_ACC_SZ_MAX );
```
Comment (topointon-jump) [inline src/flamenco/runtime/fd_executor.c:1261]: This should also include the metadata size (+sizeof(fd_account_meta_t))

File: src/flamenco/runtime/fd_runtime.c [hunk 6]
```diff
@@ -1803,7 +1810,7 @@ fd_runtime_finalize_txns_tpool( fd_exec_slot_ctx_t *         slot_ctx,
         fd_acc_mgr_view( txn_ctx->acc_mgr, txn_ctx->funk_txn, &txn_ctx->accounts[0], borrowed_account );
         memcpy( borrowed_account->pubkey->key, &txn_ctx->accounts[0], sizeof(fd_pubkey_t) );
 
-        void * borrowed_account_data = fd_valloc_malloc( txn_ctx->valloc, 8UL, fd_borrowed_account_raw_size( borrowed_account ) );
+        void * borrowed_account_data = fd_spad_alloc( txn_ctx->spad, FD_SPAD_ALIGN, FD_ACC_SZ_MAX );
```
Comment (topointon-jump) [inline src/flamenco/runtime/fd_runtime.c:1813]: metadata

**Review by topointon-jump — APPROVED** [review]

**Pull request merged**